numpy/core/src/multiarray/einsum.dispatch.c.src

```diff
@@ -0,0 +1,2071 @@
+    &@name@_sum_of_products_contig_contig_outstride0_two,
+},
+#else
+    {NULL, NULL, NULL, NULL, NULL},
+#endif
+/**end repeat**/
+}; /* End of _binary_specialization_table */
+
+static sum_of_products_fn _outstride0_specialized_table[NPY_NTYPES][4] = {
+/**begin repeat
+ * #name = bool,
+ *         byte, ubyte,
+ *         short, ushort,
+ *         int, uint,
+ *         long, ulong,
+ *         longlong, ulonglong,
+ *         float, double, longdouble,
+ *         cfloat, cdouble, clongdouble,
+ *         object, string, unicode, void,
+ *         datetime, timedelta, half#
+ * #use = 1,
+ *        1, 1,
+ *        1, 1,
+ *        1, 1,
+ *        1, 1,
+ *        1, 1,
+ *        1, 1, 1,
+ *        1, 1, 1,
+ *        0, 0, 0, 0,
+ *        0, 0, 1#
+ */
+#if @use@
+{
+    &@name@_sum_of_products_outstride0_any,
+    &@name@_sum_of_products_outstride0_one,
+    &@name@_sum_of_products_outstride0_two,
+    &@name@_sum_of_products_outstride0_three
+},
+#else
+    {NULL, NULL, NULL, NULL},
+#endif
+/**end repeat**/
+}; /* End of _outstride0_specialized_table */
+
+static sum_of_products_fn _allcontig_specialized_table[NPY_NTYPES][4] = {
+/**begin repeat
+ * #name = bool,
+ *         byte, ubyte,
+ *         short, ushort,
+ *         int, uint,
+ *         long, ulong,
+ *         longlong, ulonglong,
+ *         float, double, longdouble,
+ *         cfloat, cdouble, clongdouble,
+ *         object, string, unicode, void,
+ *         datetime, timedelta, half#
+ * #use = 1,
+ *        1, 1,
+ *        1, 1,
+ *        1, 1,
+ *        1, 1,
+ *        1, 1,
+ *        1, 1, 1,
+ *        1, 1, 1,
+ *        0, 0, 0, 0,
+ *        0, 0, 1#
+ */
+#if @use@
+{
+    &@name@_sum_of_products_contig_any,
+    &@name@_sum_of_products_contig_one,
+    &@name@_sum_of_products_contig_two,
+    &@name@_sum_of_products_contig_three
+},
+#else
+    {NULL, NULL, NULL, NULL},
+#endif
+/**end repeat**/
+}; /* End of _allcontig_specialized_table */
+
+static sum_of_products_fn _unspecialized_table[NPY_NTYPES][4] = {
+/**begin repeat
+ * #name = bool,
+ *         byte, ubyte,
+ *         short, ushort,
+ *         int, uint,
+ *         long, ulong,
+ *         longlong, ulonglong,
+ *         float, double, longdouble,
+ *         cfloat, cdouble, clongdouble,
+ *         object, string, unicode, void,
+ *         datetime, timedelta, half#
+ * #use = 1,
+ *        1, 1,
+ *        1, 1,
+ *        1, 1,
+ *        1, 1,
+ *        1, 1,
+ *        1, 1, 1,
+ *        1, 1, 1,
+ *        0, 0, 0, 0,
+ *        0, 0, 1#
+ */
+#if @use@
+{
+    &@name@_sum_of_products_any,
+    &@name@_sum_of_products_one,
+    &@name@_sum_of_products_two,
+    &@name@_sum_of_products_three
+},
+#else
+    {NULL, NULL, NULL, NULL},
+#endif
+/**end repeat**/
+}; /* End of _unnspecialized_table */
+
+NPY_NO_EXPORT sum_of_products_fn NPY_CPU_DISPATCH_CURFX(einsum_get_sum_of_products_function)
+(int nop, int type_num, npy_intp itemsize,npy_intp const *fixed_strides)
```

Comment:
The only exported part(internal exporting during the linking), the other functions, variables are **static** (private).

By assuming that requested optimizations within configuration statements are supported by the compiler, the final exported symbols on x86 will equivalent to:
 

```C
// no suffixes for the baseline
einsum_get_sum_of_products_function(int nop, int type_num, npy_intp itemsize,npy_intp const *fixed_strides);
// two underscores is used to seprate the multi-target.
einsum_get_sum_of_products_function_avx2__fma3(int nop, int type_num, npy_intp itemsize,npy_intp const *fixed_strides);
// avx512f
einsum_get_sum_of_products_function_avx512f(int nop, int type_num, npy_intp itemsize,npy_intp const *fixed_strides);
````
Note: always remember that the dispatch-able sources are compiled multiple-times according to the required optimizations to enable different instructions-sets in each compile.

Reply:
since `avx512f` implies `avx2` and `fma3`.
You have always start with the highest interested features during branching your code, same as we do in NPYV implementation. for example:
```C
#if defined(NPY_HAVE_AVX512F)
#elif defined(NPY_HAVE_AVX2)
   #if defined(NPY_HAVE_FMA3)
   #endif
#elif defined(NPY_HAVE_SSE2)
#elif defined(NPY_HAVE_VSX2)
#elif defined(NPY_HAVE_ASIMDDP)
#elif defined(NPY_HAVE_NEON)
#else
// optimizations is disabled or non-supported arch
#endif